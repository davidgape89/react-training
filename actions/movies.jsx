import axios from 'axios';

const baseUrl = 'https://reactjs-cdp.herokuapp.com';

export const startMoviesRequest = (query, searchBy) => 
  (dispatch) => axios.get(`${baseUrl}/movies?search=${query}&searchBy=${searchBy}`)
    .then(({data}) => {
      dispatch(addMovies(data.data))
    });

export const addMovies = (movies) => ({
  type: 'ADD_MOVIES',
  payload: {
    movies
  }
});