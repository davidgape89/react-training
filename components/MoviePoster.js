import React from 'react';
import PropTypes from 'prop-types';
import Link from 'next/link';

const MoviePoster = ({
  id,
  title, 
  genres,
  releaseDate,
  posterUrl,
}) => {
  const year = releaseDate.split('-')[0];
  const genre = genres.join(' & ');
  return (
    <Link href={`/film/${id}`}>
      <div className="movie-poster">
        <img src={posterUrl} />
        <div>
          <div className="movie-poster__info">
            <div className="title">{title}</div>
            <div className="year">{year}</div>
          </div>
          <div className="genre">{genre}</div>
        </div>
      </div>
    </Link>
  );
}

MoviePoster.propTypes = {
  title: PropTypes.string,
  genres: PropTypes.array,
  releaseDate: PropTypes.string,
  posterUrl: PropTypes.string
};

export default MoviePoster;