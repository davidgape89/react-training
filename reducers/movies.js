export default (state = [], {type, payload}) => {
  switch(type) {
    case 'ADD_MOVIES':
      return payload.movies;
    default:
      return state;
  }
}