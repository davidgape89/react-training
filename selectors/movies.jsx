import { createSelector } from 'reselect';

const getMovies = state => state.movies;

const getSortBy = state => state.filters.sortBy;

const getSortedMovies = createSelector(
  [getMovies, getSortBy],
  (movies, sortBy) => {
    const resultArray = [...movies];

    if (sortBy === 'release_date') {
      return resultArray.sort(
        (movie1, movie2) => {
          const timestamp1 = new Date(movie1.release_date);
          const timestamp2 = new Date(movie2.release_date);
          return timestamp1 < timestamp2 ? 1 : -1;
        },
      );
    }

    return resultArray.sort(
      (movie1, movie2) => (movie1.vote_average < movie2.vote_average ? 1 : -1),
    );
  },
);

export default getSortedMovies;
