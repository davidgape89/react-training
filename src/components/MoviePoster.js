import React from 'react';
import PropTypes from 'prop-types';

const MoviePoster = ({
  id,
  title, 
  genres,
  releaseDate,
  posterUrl,
  handleMovieSelected
}) => {
  const year = releaseDate.split('-')[0];
  const genre = genres.join(' & ');
  return (
    <div className="movie-poster" 
         onClick={() => handleMovieSelected(id)}>
      <img src={posterUrl} />
      <div>
        <div className="movie-poster__info">
          <div className="title">{title}</div>
          <div className="year">{year}</div>
        </div>
        <div className="genre">{genre}</div>
      </div>
    </div>
  );
}

MoviePoster.propTypes = {
  title: PropTypes.string,
  genres: PropTypes.array,
  releaseDate: PropTypes.string,
  posterUrl: PropTypes.string
};

export default MoviePoster;