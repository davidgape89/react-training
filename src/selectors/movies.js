export default (movies, sortBy) => {
  let resultArray = [...movies];
  if(sortBy === 'release_date') {
    return resultArray.sort(
      (movie1, movie2) => {
        const timestamp1 = new Date(movie1.release_date);
        const timestamp2 = new Date(movie2.release_date);
        return timestamp1 < timestamp2? 1: -1;
      });
  } else {
    return resultArray.sort(
      (movie1, movie2) => {
        return movie1.vote_average < movie2.vote_average? 1: -1;
      }
    );
  }
}