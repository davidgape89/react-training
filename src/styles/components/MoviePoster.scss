@import '../styles/variables';

.movie-poster {
  font-size: 10px;

  img {
    width: 260px;
  }

  > div {
    padding: $small-size;
  }
}

.movie-poster__info {
  align-items: center;
  display: flex;
  justify-content: space-between;
  margin-bottom: 4px;

  .title {
    font-size: 16px;
  }

  .year {
    border: 1px solid grey;
    border-radius: 2px;
    padding: 3px;
  }
}