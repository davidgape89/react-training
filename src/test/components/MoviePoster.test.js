import React from 'react';
import {shallow} from 'enzyme';

import movies from '../../mocks/movie';
import MoviePoster from '../../components/MoviePoster';

describe('MoviePoster -', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<MoviePoster title={movies[0].title}
                                   genres={movies[0].genres}
                                   release_date={movies[0].release_date}
                                   poster_path={movies[0].poster_url} />);
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});