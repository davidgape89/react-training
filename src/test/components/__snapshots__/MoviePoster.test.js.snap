// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieList - renders correctly 1`] = `
<div
  className="movie-poster"
  onClick={[Function]}
>
  <img />
  <div>
    <div
      className="movie-poster__info"
    >
      <div
        className="title"
      >
        Fifty Shades Freed
      </div>
      <div
        className="year"
      >
        2018
      </div>
    </div>
    <div
      className="genre"
    >
      Drama & Romance
    </div>
  </div>
</div>
`;
